www/js/metrics/MetricsTab.tsx

```diff
@@ -16,53 +15,101 @@ import DailyActiveMinutesCard from './DailyActiveMinutesCard';
   };
   if (type == 'user') return getMetrics('timestamp', query);
   return getAggregateData('result/metrics/timestamp', query, serverConnConfig);
 }
 
-function getLastTwoWeeksDtRange() {
-  const now = DateTime.now().startOf('day');
-  const start = now.minus({ days: 15 });
-  const end = now.minus({ days: 1 });
-  return [start, end];
-}
-
 const MetricsTab = () => {
   const appConfig = useAppConfig();
   const { colors } = useTheme();
   const { t } = useTranslation();
   const { getFormattedSpeed, speedSuffix, getFormattedDistance, distanceSuffix } =
     useImperialConfig();
+  const {
+    dateRange,
+    setDateRange,
+    timelineMap,
+    timelineLabelMap,
+    timelineIsLoading,
+    refreshTimeline,
+    loadMoreDays,
+  } = useContext(TimelineContext);
 
-  const [dateRange, setDateRange] = useState<DateTime[]>(getLastTwoWeeksDtRange);
   const [aggMetrics, setAggMetrics] = useState<MetricsData | undefined>(undefined);
-  const [userMetrics, setUserMetrics] = useState<MetricsData | undefined>(undefined);
 
+  // user metrics are computed on the phone from the timeline data
+  const userMetrics = useMemo(() => {
+    console.time('MetricsTab: generate_summaries');
+    if (!timelineMap) return;
+    console.time('MetricsTab: timelineMap.values()');
+    const timelineValues = [...timelineMap.values()];
+    console.timeEnd('MetricsTab: timelineMap.values()');
+    const result = MetricsSummaries.generate_summaries(
+      METRIC_LIST,
+      timelineValues,
+      timelineLabelMap,
+    ) as MetricsData;
+    console.timeEnd('MetricsTab: generate_summaries');
+    return result;
+  }, [timelineMap]);
+
+  // at least 2 weeks of timeline data should be loaded for the user metrics
   useEffect(() => {
     if (!appConfig?.server) return;
-    loadMetricsForPopulation('user', dateRange);
-    loadMetricsForPopulation('aggregate', dateRange);
-  }, [dateRange, appConfig?.server]);
+    const dateRangeDays = isoDatesDifference(...dateRange);
+
+    // this tab uses the last 2 weeks of data; if we need more, we should fetch it
+    if (dateRangeDays < 14) {
+      if (timelineIsLoading) {
+        logDebug('MetricsTab: timeline is still loading, not loading more days yet');
+      } else {
+        logDebug('MetricsTab: loading more days');
+        loadMoreDays('past', 14 - dateRangeDays);
+      }
+    } else {
+      logDebug('MetricsTab: date range >= 14 days, not loading more days');
+    }
+  }, [dateRange, timelineIsLoading, appConfig?.server]);
+
+  // aggregate metrics fetched from the server whenever the date range is set
+  useEffect(() => {
+    logDebug('MetricsTab: dateRange updated to ' + JSON.stringify(dateRange));
+    const dateRangeDays = isoDatesDifference(...dateRange);
+    if (dateRangeDays < 14) {
+      logDebug('MetricsTab: date range < 14 days, not loading aggregate metrics yet');
+    } else {
+      loadMetricsForPopulation('aggregate', dateRange);
+    }
+  }, [dateRange]);
```

Comment:
How about defining a constant variable for the number 14? If we later decide to display only 7 days or extend it to 21 days, we can easily modify just one variable

Reply:
Good idea, I will make this change